conan/tools/qbs/generic.py

```diff
@@ -0,0 +1,296 @@
+        self._parse = self._add_linker_flag
+
+    def _add_linker_flag(self, token):
+        self.linker_flags.append(token)
+        self._parse = self._detect_wl_or_detect_comma_or_add_driver_linker_flag
+
+
+def _flags_from_env():
+    flags_from_env = {}
+    if tools.get_env('ASFLAGS'):
+        flags_from_env['cpp.assemblerFlags'] = '%s' % (
+            _env_var_to_list(tools.get_env('ASFLAGS')))
+    if tools.get_env('CFLAGS'):
+        flags_from_env['cpp.cFlags'] = '%s' % (
+            _env_var_to_list(tools.get_env('CFLAGS')))
+    if tools.get_env('CPPFLAGS'):
+        flags_from_env['cpp.cppFlags'] = '%s' % (
+            _env_var_to_list(tools.get_env('CPPFLAGS')))
+    if tools.get_env('CXXFLAGS'):
+        flags_from_env['cpp.cxxFlags'] = '%s' % (
+            _env_var_to_list(tools.get_env('CXXFLAGS')))
+    if tools.get_env('LDFLAGS'):
+        parser = LinkerFlagsParser(_env_var_to_list(tools.get_env('LDFLAGS')))
+        flags_from_env['cpp.linkerFlags'] = str(parser.linker_flags)
+        flags_from_env['cpp.driverLinkerFlags'] = str(
+            parser.driver_linker_flags)
+    return flags_from_env
+
+
+class QbsGenericToolchain(object):
+    filename = 'conan_toolchain.qbs'
+
+    _template_toolchain = textwrap.dedent('''\
+        import qbs
+
+        Project {
+            Profile {
+                name: "conan_toolchain_profile"
+
+                /* detected via qbs-setup-toolchains */
+                {%- for key, value in _profile_values_from_setup.items() %}
+                {{ key }}: {{ value }}
+                {%- endfor %}
+
+                /* deduced from environment */
+                {%- for key, value in _profile_values_from_env.items() %}
+                {{ key }}: {{ value }}
+                {%- endfor %}
+                {%- if sysroot %}
+                qbs.sysroot: "{{ sysroot }}"
+                {%- endif %}
+
+                /* conan settings */
+                {%- if build_variant %}
+                qbs.buildVariant: "{{ build_variant }}"
+                {%- endif %}
+                {%- if architecture %}
+                qbs.architecture: "{{ architecture }}"
+                {%- endif %}
+                {%- if optimization %}
+                qbs.optimization: "{{ optimization }}"
+                {%- endif %}
+                {%- if cxx_language_version %}
+                cpp.cxxLanguageVersion: "{{ cxx_language_version }}"
+                {%- endif %}
+
+                /* package options */
+                {%- if position_independent_code %}
+                cpp.positionIndependentCode: {{ position_independent_code }}
+                {%- endif %}
+            }
+        }
+        ''')
+
+    def __init__(self, conanfile):
+        _check_for_compiler(conanfile)
+        self._conanfile = conanfile
+        _setup_toolchains(conanfile)
+        self._profile_values_from_setup = (
+            _read_qbs_toolchain_from_config(conanfile))
```

Comment:
I am a bit confused about this round-trip conversion, but probably it is because I don't know Qbs enough. If the config data is already present in a qbs format file, why is it necessary to parse it, and translate it to a conan-toolchain format one? Couldn't it be used directly from that original qbs origin?

Reply:
I use the qbs command `qbs-setup-toolchains` here because it contains some logic on how to detect compilers and create a qbs-profile out of it. I did not want to mirror the logic here. The problem is that this tool creates qbs-settings which are by default user/system wide and are stored in a `.ini` file containing binary Qt data. So parsing this settings file cannot be easily be done by in python. This is why I use another qbs command `qbs-config` to read the values in a easy to parse format.